mri/reconstruct/fourier.py

```diff
@@ -208,3 +208,139 @@ def adj_op(self, x):
         """
         self.plan.f = x
         return np.copy(self.plan.adjoint()) / np.sqrt(self.plan.M)
+
+
+class Stacked3D:
+    """"  3-D non uniform Fast Fourrier Transform class,
+        fast implementation for Stacked samples
+    Attributes
+    ----------
+    samples: np.ndarray
+        the mask samples in the Fourier domain.
+    shape: tuple of int
+        shape of the image (necessarly a square/cubic matrix).
+    fft_type: 'NFFT' | 'NUFFT' default 'NFFT'
+        What FFT version to be used for 2D non uniform FFT
+    platform: string, 'cpu', 'multi-cpu' or 'gpu' default 'gpu'
+        string indicating which hardware platform will be used to compute the
+        NUFFT. works only if fft_type=='NUFFT'
+    """
+
+    def __init__(self, samples, shape, fft_type='NFFT', platform='gpu'):
+        """ Init function for Stacked3D class
+        Parameters
+        ----------
+        samples: np.ndarray
+            the mask samples in the Fourier domain.
+        shape: tuple of int
+            shape of the image (necessarly a square/cubic matrix).
+        fft_type: 'NFFT' | 'NUFFT' default 'NFFT'
+            What FFT version to be used for 2D non uniform FFT
+        platform: string, 'cpu', 'multi-cpu' or 'gpu' default 'gpu'
+            string indicating which hardware platform will be used to
+            compute the NUFFT. works only if fft_type=='NUFFT'
+        """
+        self.num_slices = shape[2]
+        self.shape = shape
+        # Sort the incoming data based on Z, Y then X coordinates
+        # This is done for easier stacking
+        self.sort_pos = np.lexsort(tuple(samples[:, i]
+                                         for i in np.arange(3)))
+        samples = samples[self.sort_pos]
+        plane_samples, self.z_samples = self.get_stacks(samples)
+        if fft_type == 'NUFFT':
+            self.FT = NUFFT(samples=plane_samples, shape=shape[0:2],
+                            platform=platform)
+        elif fft_type == 'NFFT':
+            self.FT = NFFT(samples=plane_samples, shape=shape[0:2])
+
+    def get_stacks(self, samples):
+        """ Function that converts an incoming 3D kspace samples
+            and converts to stacks of 2D. This function also checks for
+            any issues incoming k-space pattern and if the stack property
```

Comment:
typo I think: 'any issues of the incoming k-space'

Reply:
yes, sorry